core/string/string_name.cpp

```diff
@@ -244,8 +253,8 @@ StringName::StringName(const char *p_name, bool p_static) {
 
 	_data = memnew(_Data);
```

Comment:
I'm not sure how comfortable I am with memory allocations being done while holding a global spinlock. (And same concern for the `memdelete` calls as well.

If there's some way we can put have a sort of optimistic memory pool / some way to do these memory ops while not holding a lock, it would be nice.

The `memdelete` should definitely be very possible to defer until after we have released the spinlock, since we can hold a local pointer set to null and then check afterwards if it's not null and delete it.

the memnew I'm not sure: you don't want to optimistically allocate beforehand. Perhaps there could be a "spare"  instance and it can be replenished before acquiring the spinlock if it's null, and freed after releasing the spinlock if someone else replenished it first. So that way you can safely do one allocation without hitting the memory allocator while locked.